### PR TITLE
Internal Changes: Refactor VisitorAPI to event based interaction

### PR DESCRIPTION
**Pull request by Helfull**
With this PR the `VisitorAPI` or `GardenVisitorFeature` is moved to an event based system.

There is one central "Listener" for the base events from MC/Forge and spawns the required Visitor events which then will be handled by the required features.

I had to also change the `RenderUtils` as i didnt want to expose the `RenderWorldLastEvent` through the `VisitorRenderEvent` and since it was not used anyway it was safe to be changed, this can be seen in this commit https://github.com/hannibal002/SkyHanni/pull/561/commits/b453b934b56efc214c874ae6d2f82f50d058564d

This PR should be merged/checked after this pr https://github.com/hannibal002/SkyHanni/pull/555

### CODE REVIEW

**Comment by hannibal002** [comment]
I don't like having two independent changes in the same pr :/

image

**Comment by Helfull** [comment]
> I don't like having two independent changes in the same pr :/
> 
> image

As i describe, it was this, or expose the `RenderWorldLastEvent`

I can pull those changes out into its own PR, which would then be blocking this one

**Comment by hannibal002** [comment]
I don't see the harm in doing so.
See https://github.com/hannibal002/SkyHanni/blob/39adca7678bb675e6917b35c1de699c6a0bd55dc/src/main/java/at/hannibal2/skyhanni/features/bazaar/BazaarApi.kt#L62

**Comment by hannibal002** [comment]
Or a third approach would be to not expose the render event directly but provide the drawString method inside the visitor event that then calls the render event internally.

**Comment by hannibal002** [comment]
Any way, I don't like the idea of being able to run render calls outside their intended scope. This is a part of bug prevention framework design in my mind.

**Comment by Helfull** [comment]
i will expose the event, over the third option, which i was thinking about as well, but it felt wrong.

File: src/main/java/at/hannibal2/skyhanni/features/garden/visitor/VisitorTooltipParser.kt
Comment (Helfull) [inline src/main/java/at/hannibal2/skyhanni/features/garden/visitor/VisitorTooltipParser.kt:1]: This one was supposed to be used in the `readTooltip` method to make it easier to read.

But i threw most of it away, after i started working on the event based refactor. I will add those changes back in the next change set after the events are done.

**Comment by Helfull** [comment]
Converting to draft, as the last changes on beta broke a lot here with conflicts. Will add them

**Comment by Helfull** [comment]
I dont know why these tests fail now, need to check that later

**Comment by Helfull** [comment]
Okay, the cause for this is the definition of `fun SkyBlockTime.formatted(): String` this seems not to be resolvable with mockk

**Comment by Helfull** [comment]
@hannibal002 maybe we should stop declaring extensions inside other classes, if they dont need any internals of that class to work.

This would remove the size of those classes, in additions we could create a `utils.extensions` package which contains all of those extensions which are created, this way it is easier to identify if a function is a member or an extension we created.

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**